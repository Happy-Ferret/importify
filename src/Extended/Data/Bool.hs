-- | Extended utilities for 'Bool' type.

module Extended.Data.Bool
       ( (==>)
       ) where

import           Universum

-- | Logical implication.
(==>) :: Bool -> Bool -> Bool
False ==> _ = True
True  ==> x = x
infixr 4 ==>
