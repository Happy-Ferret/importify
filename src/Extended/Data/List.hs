-- | This module contains additional utility functions for list.

module Extended.Data.List
       ( removeAt
       , removeAtMultiple
       ) where

import           Universum

-- | Removes element from list by given index. If there's
-- no element at such index then list returns unchanged.
removeAt :: Int -> [a] -> [a]
removeAt i l | i < 0 = l
removeAt i l = case after of
                     []       -> l
                     (_:rest) -> before ++ rest
  where (before, after) = splitAt i l

-- | Like 'removeAt' but takes list of indices to remove.
removeAtMultiple :: [Int] -> [a] -> [a]
removeAtMultiple indices = map snd
                         . filter ((`notElem` indices) . fst)
                         . zip [0..]
