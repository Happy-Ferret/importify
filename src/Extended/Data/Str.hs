-- | Utilities to work with textual types.

module Extended.Data.Str
       ( -- * Wrapping
         charWrap
       , wordWrap
       ) where

import           Universum

import qualified Data.Text as T

-- | Wraps given string into lines non exceeding given length
-- splitting by chars.
charWrap :: Int -> String -> Text
charWrap n = unlines . T.chunksOf n . toText

-- | Wraps given string into lines non exceeding given length
-- splitting by words.
wordWrap :: Int -> String -> Text
wordWrap n = unlines . map unwords . go 0 [] . words . toText
  where
    go :: Int -> [Text] -> [Text] -> [[Text]]
    go _   row    []  = [row]
    go acc row (w:ws) = let newAcc = acc + T.length w
                        in if newAcc > n
                           then row : go 0 [] ws
                           else go newAcc (row ++ [w]) ws
