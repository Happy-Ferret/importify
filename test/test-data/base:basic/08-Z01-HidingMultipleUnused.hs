module HidingPartiallyUnused where

import Data.List hiding (lookup, sort)
import GHC.OldList (sort)

foo = sort
