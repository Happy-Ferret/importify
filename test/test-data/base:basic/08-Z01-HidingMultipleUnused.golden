module HidingPartiallyUnused where

import GHC.OldList (sort)

foo = sort
